gqueue.h

```diff
@@ -9,19 +9,19 @@
 struct generic_queue {
   volatile void *head;
   volatile void *tail;
-  int32_t item_size;
-  int32_t len;
-  int32_t max_capacity;
+  volatile void *end;
+  unsigned int item_size;
+  volatile unsigned int len;
+  unsigned int max_capacity;
   volatile uint8_t memory[0];
 };
 
 #define DECLARE_QUEUE(name, element_type, max_size)	\
   static struct {					\
     struct generic_queue gq;				\
-    element_type __elements[max_size];			\
+    element_type __elements[max_size + 1];              \
```

Comment:
Double underscores are reserved for use by standard headers or the compiler and shouldn't be used. It can cause the program's behavior to be undefined.

Maybe I'm wrong, but the way I understand "always reserved for any use" is that it is reserved for code in standard headers and the compiler itself.

Here is an [answer](http://stackoverflow.com/questions/25090635/use-and-in-c-programs) that I found about it on-line. 

> Here's what the C standard says (section 7.1.3):
> 
> All identifiers that begin with an underscore and either an uppercase letter or another underscore are always reserved for any use.
> All identifiers that begin with an underscore are always reserved for use as identifiers with file scope in both the ordinary and tag name spaces.
> (The section goes on to list specific identifiers and sets of identifiers reserved by certain standard headers.)
> 
> What this means is that for example, the implementation (either the compiler or a standard header) can use the name __FOO for anything it likes. If you define that identifier in your own code, your program's behavior is undefined. If you're "lucky", you'll be using an implementation that doesn't happen to define it, and your program will work as expected